File: c++/0155-min-stack/min_stack.cpp
#include <stack>
#include <utility>

using namespace std;

class MinStack {
    stack<int> s;
    stack<pair<int, int>> ms;

public:
    MinStack() {}

    void push(int val) {
        s.push(val);

        if (ms.empty() || val < ms.top().first) {
            ms.push({val, 1});
        } else if (val == ms.top().first) {
            ++ms.top().second;
        }
    }

    void pop() {
        if (s.top() == ms.top().first) {
            if (--ms.top().second == 0) {
                ms.pop();
            }
        }

        s.pop();
    }

    int top() {
        return s.top();
    }

    int getMin() {
        return ms.top().first;
    }
};

File: java/0155-min-stack/MinStack.java
import java.util.Stack;

public class MinStack {
    private final Stack<Integer> stack = new Stack<>();
    private final Stack<Integer> minStack = new Stack<>();

    public MinStack() {
    }

    public void push(int val) {
        stack.push(val);
        minStack.push(minStack.isEmpty() ? val : Math.min(val, minStack.peek()));
    }

    public void pop() {
        stack.pop();
        minStack.pop();
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}
